infra/src/main/kotlin/fi/fta/geoviite/infra/linking/switches/SwitchLinkingService.kt

```diff
@@ -1245,3 +1088,48 @@ fun matchSamplingGridToQueryPoints(
         },
     )
 }
+
+class SuggestedSwitchWithOriginallyLinkedTracks(
+    val suggestedSwitch: SuggestedSwitch,
+    val originallyLinkedTracks: Map<IntId<LocationTrack>, Pair<LocationTrack, LayoutAlignment>>,
+) {
+
+    // need to use these as map keys, but avoid very expensive hashCode()s on all those alignments
+    // with their geometries
```

Comment:
Aika pitkä asia selittää tämä optimointi...

Vaihteiden live-linkityksessä (joka saadaan joskus toiminaan frontillakin asti...) yksi tärkeähkö optimointi on, että aivan vierekkäisillä pikseleillä voi hyvinkin olla, että pyydettäessä niiden kohdalle vaihde-ehdotusta ainakin osassa laskentoja käydään läpi itse asiassa ihan samoja tietoja, ja siksi meillä on PointAssociation-luokka, joka yhdistelee näitä laskutoimituksia ja valvoo samalla, mihin pikseliin ne alkuaan liittyivät.

Yhdistely toimii käyttämällä tietoja HashMapin avaimina, ja varsinainen vaihde-ehdotus on yksi näistä mahdollisesti yhdistettävistä tiedoista (ts. vierekkäisillä pikseleillä saattaa olla sama vaihde-ehdotus). Mutta toisaalta samasta paikasta sattuu olemaan optimoinnin suhteen hyötyä tuoda myös tieto siitä, millä raiteilla tämä vaihde oli alkuaan kiinni, ja helpoiten ihan varsinaisina raide- ja alignment-olioina.

Mutta jos niin tekee, niin sitten se HashMapin avainten laskenta käy läpi noiden linjojen segmenttien jokaisen pisteen saadakseen aikaan niiden hash-arvot. (Ehkä voisi pitää parempana ratkaisuna toteuttaa vaan LayoutAlignmentille hashCode-metodin, joka ei katso sen segmenttejä? En tiedä.) Parista mahdollisesta ratkaisusta pidin tätä järkevimpänä tapana estää outouksia leviämästä ympäri koodia, tosin en kyllä kovin tarkkaan ajatellut, olisiko parempaakin mahdollisuutta.

Reply:
Joo kommentin lukiessa tuli sama mieleen että meillä voisi olla equals ja hashcode toteutukset kaikille layout-käsitteille + alignmentille + segmentille perustuen niiden rowversioon (yhden version tiedot ei koskaan muutu), ettei tarvitse tehdä tällaisia kikkoja. Vastaavasti vielä segmentgeometry-oliolle sen voisi tehdä vain id:llä koska niitä ei koskaan muokata ja haluttaessa myös planeille jne voi tehdä saman rowversiolla. 

Pari ongelmaa tuossa toki on:
- Tuo ei silloin toimi niille olioille joita ei ole vielä tallennettu kantaan, mutta sen näkee version puuttumisesta ja silloin voi vaikka defaultata superiin
- Jos jotain data-oliota muokataan kotlin-koodissa esim dataclassin copy-metodilla, noista pitäisi jotenkin onnistua tajuamaan että tämä data ei enää vastaa versiota, joten pitäisi taas defaultata johonkin monimutkaisempaan. Sikäli kun muokkaukset on olion omilla metodeilla, tuon voinee tehdä jollain dirty-flagilla tms, mutta milläpä napata default copy-funkkari l:thinking: